File: 卡码/代码随想录/1/11-图论/Js/dfs/LeetCode_0827.js
//827. 最大人工岛
const largestIsland = function(grid) {
  const dir=[[0,1],[1,0],[0,-1],[-1,0]];
  const n=grid.length,m=grid[0].length;
  let cnt=0;
  const dfs=(x,y,mark)=>{
    grid[x][y]=mark;
    cnt++;
    for(let i=0;i<4;i++){
      const nextX=x+dir[i][0];
      const nextY=y+dir[i][1];
      if(nextX<0||nextX>n-1||nextY<0||nextY>m-1) continue;
      if(grid[nextX][nextY]==0||grid[nextX][nextY]>1) continue;
      dfs(nextX,nextY,mark);
    }
  }
  let mark=2;
  let isAllLand=true;
  const areas=new Map();
  for(let i=0;i<n;i++){
    for(let j=0;j<m;j++){
      if(grid[i][j]==0) isAllLand=false;
      if(grid[i][j]==1){
        cnt=0;
        dfs(i,j,mark);
        areas.set(mark,cnt);
        mark++;
      }
    }
  }
  if(isAllLand) return n*m;
  const visited=new Set();
  let res=0;
  for(let i=0;i<n;i++){
    for(let j=0;j<m;j++){
      if(grid[i][j]==0){
        let count=1;
        visited.clear();
        for(let k=0;k<4;k++){
          const x=i+dir[k][0];
          const y=j+dir[k][1];
          if(x<0||x>n-1||y<0||y>m-1) continue;
          if(grid[x][y]>1){
            if(visited.has(grid[x][y])) continue;
            count+=areas.get(grid[x][y]);
            visited.add(grid[x][y]);
          }
        }
        res=Math.max(res,count);
      }
    }
  }
  return res;
};

File: 卡码/代码随想录/1/11-图论/Go/dfs/LeetCode_0827.go
// 827. 最大人工岛
package main

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

func largestIsland(grid [][]int) int {
	n, m := len(grid), len(grid[0])
	dir := [2][4]int{[4]int{0, 1, 0, -1}, [4]int{1, 0, -1, 0}}
	cnt := 0
	var dfs func(x, y, mark int)
	dfs = func(x, y, mark int) {
		cnt++
		grid[x][y] = mark
		for i := 0; i < 4; i++ {
			nextX, nextY := x+dir[0][i], y+dir[1][i]
			if nextX < 0 || nextX > n-1 || nextY < 0 || nextY > m-1 {
				continue
			}
			if grid[nextX][nextY] == 0 || grid[nextX][nextY] > 1 {
				continue
			}
			dfs(nextX, nextY, mark)
		}
	}
	mark := 2
	isAllLand := true
	areas := make(map[int]int)
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == 0 {
				isAllLand = false
			}
			if grid[i][j] == 1 {
				cnt = 0
				dfs(i, j, mark)
				areas[mark] = cnt
				mark++
			}
		}
	}
	if isAllLand {
		return n * m
	}
	res := 0
	visited := make(map[int]bool)
	for i := 0; i < n; i++ {
		for j := 0; j < m; j++ {
			if grid[i][j] == 0 {
				cnt := 1
				visited = make(map[int]bool)
				for k := 0; k < 4; k++ {
					nextX, nextY := i+dir[0][k], j+dir[1][k]
					if nextX < 0 || nextX > n-1 || nextY < 0 || nextY > m-1 {
						continue
					}
					if _, ok := visited[grid[nextX][nextY]]; ok {
						continue
					}
					visited[grid[nextX][nextY]] = true
					cnt += areas[grid[nextX][nextY]]
				}
				res = max(res, cnt)
			}
		}
	}
	return res
}
